File: packages/urdu-js/src/tokenization/tokenize.ts
import type { Token } from '../types.js'
import { classifyChar } from '../analysis/char-class.js'

function classifyTokenType(text: string): Token['type'] {
  if (!text.trim()) return 'whitespace'
  const chars = [...text]
  let urdu = 0, latin = 0, num = 0
  for (const ch of chars) {
    const cls = classifyChar(ch)
    if (cls === 'urdu-letter') urdu++
    else if (cls === 'latin') latin++
    else if (cls === 'numeral') num++
  }
  const total = chars.length
  if (num === total) return 'numeral'
  if (urdu / total >= 0.8) return 'urdu-word'
  if (latin / total >= 0.8) return 'latin-word'
  if (urdu > 0 && latin > 0) return 'mixed'
  return 'urdu-word'
}

export function tokenize(text: string): Token[] {
  if (!text) return []
  return text
    .split(/([\u0020\u00a0\u200b\t\n\r\v]+)/)
    .filter(p => p.length > 0)
    .map(p => ({ text: p, type: classifyTokenType(p) }))
}

File: packages/urdu-dart/lib/src/tokenization/tokenize.dart
import '../types.dart';
import '../analysis/char_class.dart';

String _classifyTokenType(String text) {
  if (text.trim().isEmpty) return 'whitespace';
  final runes = text.runes.toList();
  int urdu = 0, latin = 0, num = 0;
  for (final r in runes) {
    final ch = String.fromCharCode(r);
    final cls = classifyChar(ch);
    if (cls == CharClass.urduLetter) {
      urdu++;
    } else if (cls == CharClass.latin) {
      latin++;
    } else if (cls == CharClass.numeral) {
      num++;
    }
  }
  final total = runes.length;
  if (num == total) return 'numeral';
  if (urdu / total >= 0.8) return 'urdu-word';
  if (latin / total >= 0.8) return 'latin-word';
  if (urdu > 0 && latin > 0) return 'mixed';
  return 'urdu-word';
}

/// Tokenizes Urdu text into a list of tokens.
///
/// Splits on whitespace characters (space, NBSP, ZWSP, tab, newline).
/// Each token includes its text and type classification.
List<Token> tokenize(String text) {
  if (text.isEmpty) return [];
  final parts = text.split(RegExp(r'([\u0020\u00a0\u200b\t\n\r\v]+)'));
  return parts
      .where((p) => p.isNotEmpty)
      .map((p) => Token(text: p, type: _classifyTokenType(p)))
      .toList();
}
